sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CTESubstitution.scala

```diff
@@ -211,8 +212,33 @@ object CTESubstitution extends Rule[LogicalPlan] {
       } else {
         // A CTE definition might contain an inner CTE that has a higher priority, so traverse and
         // substitute CTE defined in `relation` first.
+        // NOTE: we must call `traverseAndSubstituteCTE` before `substituteCTE`, as the relations
+        // in the inner CTE have higher priority over the relations in the outer CTE when resolving
+        // inner CTE relations. For example:
+        // WITH t1 AS (SELECT 1)
+        // t2 AS (
+        //   WITH t1 AS (SELECT 2)
+        //   WITH t3 AS (SELECT * FROM t1)
+        // )
+        // t3 should resolve the t1 to `SELECT 2` instead of `SELECT 1`.
         traverseAndSubstituteCTE(relation, isCommand, cteDefs)._1
       }
+
+      if (cteDefs.length > lastCTEDefCount) {
```

Comment:
Please consider https://github.com/apache/spark/pull/36146 as an alternative to substituting and changing accumulated `cteDefs` so far.

Reply:
Hi @peter-toth , this PR contains all the CTE bug fixes we have found so far internally. Can you rebase #36146 after this one gets merged if you think your fix is cleaner? thanks!

Reply:
I can rebase my https://github.com/apache/spark/pull/36146, no problem with that.

But I'm more concerned about https://github.com/apache/spark/pull/32298. This PR seems to contain a mix of improvements and bugfixes and changes a lot in CTE handling and conflicts with my PR. As mine is on the 3.3 whitelist can we merge that first and rebase this on that?

Reply:
This PR will be merged to 3.2 because of the fixes for the correctness bugs and performance regressions. It doesn't really improve the performance compared to Spark 3.1.

Reply:
But can we merge it after https://github.com/apache/spark/pull/32298?

Reply:
https://github.com/apache/spark/pull/32298 won't go to 3.2, right? If we really need a different CTE handling in master/3.3, we should still merge this PR first and make a followup PR to change CTE in master/3.3

Reply:
Oh I though you just made a typo and want to merge it 3.3+ only and maybe backport the bugfix parts to 3.2...